Support threading in incompressible Navier Stokes

lindsayad opened the pull request:
Force `INSADObjectTracker` to take names instead of pointers to threaded data.

Closes #15713

lindsayad committed:
Add a name member to MaterialProperty

Refs #15713

lindsayad committed:
Force INSADObjectTracker to take names instead of pointers

This enables consistency checking in threaded contexts

Closes #15713

lindsayad committed:
Add some threaded tests

Refs #15713

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/578607/) on 4ef9aba wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/15714/site)

This comment will be updated on new commits.

lindsayad committed:
Specialize for MooseEnum to avoid deprecation message

rwcarlsen reviewed (dismissed):
Looks fine to me.  I'll give @YaqiWang a chance to chime in before merging.

YaqiWang commented on framework/include/materials/MaterialProperty.h:
```diff
@@ -180,6 +182,21 @@ class MaterialPropertyBase : public PropertyValue
    */
   virtual void load(std::istream & stream) override;
 
+  void name(const MaterialPropertyName & name_in)
```
typically `name()` suggested a getter but this is a setter, so suggest to rename it to `setName`.

YaqiWang commented on modules/navier_stokes/src/kernels/INSADBoussinesqBodyForce.C:
```diff
@@ -39,9 +39,18 @@ INSADBoussinesqBodyForce::INSADBoussinesqBodyForce(const InputParameters & param
+  obj_tracker.set("alpha", getADMaterialProperty<Real>("alpha_name").name());
+  obj_tracker.set("ref_temp", getMaterialProperty<Real>("ref_temp").name());
+
+  mooseAssert(getParam<std::vector<VariableName>>("temperature").size() == 1,
```
`coupledComponents` can do the same as `getParam<std::vector<VariableName>>("temperature").size()`.

YaqiWang replied:
I feel it is ok to change this to a `mooseError`.

YaqiWang commented on modules/navier_stokes/src/kernels/INSADEnergySource.C:
```diff
@@ -49,9 +49,13 @@ INSADEnergySource::INSADEnergySource(const InputParameters & parameters)
   if (has_coupled)
-    obj_tracker.set("heat_source_var", &adCoupledValue("source_variable"));
+  {
+    mooseAssert(getParam<std::vector<VariableName>>("source_variable").size() == 1,
```
same here.

YaqiWang commented on modules/navier_stokes/src/kernels/INSADEnergySource.C:
```diff
@@ -49,9 +49,13 @@ INSADEnergySource::INSADEnergySource(const InputParameters & parameters)
+  {
+    mooseAssert(getParam<std::vector<VariableName>>("source_variable").size() == 1,
+                "Only expect one VariableName for the heat source var");
+    obj_tracker.set("heat_source_var", getParam<std::vector<VariableName>>("source_variable")[0]);
```
same here.

YaqiWang commented on modules/navier_stokes/src/kernels/INSADBoussinesqBodyForce.C:
```diff
@@ -39,9 +39,18 @@ INSADBoussinesqBodyForce::INSADBoussinesqBodyForce(const InputParameters & param
+  mooseAssert(getParam<std::vector<VariableName>>("temperature").size() == 1,
+              "Only expect one variable name for temperature");
+
+  obj_tracker.set("temperature", getParam<std::vector<VariableName>>("temperature")[0]);
```
`getVar("source_variable", 0)->name()` looks better.

YaqiWang commented on modules/navier_stokes/src/kernels/INSADMomentumCoupledForce.C:
```diff
@@ -49,19 +49,10 @@ INSADMomentumCoupledForce::INSADMomentumCoupledForce(const InputParameters & par
-      coupled_vars.push_back(&adCoupledVectorValue("coupled_vector_var", i));
-    obj_tracker.set("coupled_force_var", coupled_vars);
-  }
+    obj_tracker.set("coupled_force_var", getParam<std::vector<VariableName>>("coupled_vector_var"));
```
Not going to list all.

lindsayad replied:
I feel like this one is fine right? I don't see any `Coupleable` APIs that return the whole vector of coupled variables given a name

YaqiWang commented on modules/navier_stokes/test/tests/ins/boussinesq/tests:
```diff
@@ -7,6 +7,14 @@
+    input = boussinesq_square.i
+    exodiff = boussinesq_square_out.e
+    requirement = 'The system shall be able to solve mass, momentum, and energy incompressible Navier-Stokes equations with multiple threads.'
+    min_threads = 2
```
I guess this is a short term solution before we add the multi-thread test target.

lindsayad replied:
Yes although it would obviously work still even after we added the threading target. We would just have duplicate tests when running that target.

YaqiWang commented on modules/navier_stokes/src/userobjects/INSADObjectTracker.C:
```diff
@@ -77,3 +77,10 @@ addAmbientConvectionParams(InputParameters & params)
+
+template <>
+bool
+INSADObjectTracker::notEqual(const MooseEnum & val1, const MooseEnum & val2)
```
Is it necessary to create this function? The diff indicates that we only need to change one place, which we can just directly use `compareCurrent`.

lindsayad replied:
If I don't have this function then I have to do a whole specialization for `INSADObjectTracker::set` for `MooseEnum`, which is fine if that's what you prefer. Just more duplicate code (although only 10 lines)

YaqiWang reviewed (changes requested):
All comments are minor. One question is that will we add the multi-thread target? If so, one commit will need to be reverted in the future.

lindsayad committed:
Address review

lindsayad committed:
Use std::string for variable names

getVar->name() returns a std::string unfortunately, so that necessitates
a change in the INSADObjectTracker parameter type

lindsayad commented:
Ok, I need a re-approval please :smile:
